# Data Science Capstone - Coursera
# Sales calls for Brewery Equipment Sales visit to Austin, TX

## Introduction

### I am a sales representative for company a that sells manufacturing equipment to beer breweries. I have heard there are a lot of breweries in the Austin, TX, USA metro area. I would like to pull location data for all the breweries in the Austin area, find the zip code with the most amount of breweries, and then book the hotel with the highest rating in that zip. 

### the data I will pull will be publicly available zip code data. This will help me make a map of Austin. I will then pull foursquare data of brewery venues in austin and match those to the zip codes.  I will cluster the breweries by zip code to find the zipcodes with the most breweries. I will then utilize the best clusters to find hotels with the highest ranking by pull the venue rating from foursquare. 

# Data - let's pull the zip code and foursquare venue data

In [1]:
# Pull zip code data from public source
import pandas as pd

list = pd.read_csv("https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/download/?format=csv&timezone=America/New_York&use_labels_for_header=true",sep=';')
list.head()

Zip        City State   Latitude   Longitude  Timezone  \
0  67553  Liebenthal    KS  38.654948  -99.320620        -6   
1  85743      Tucson    AZ  32.335122 -111.148880        -7   
2  75016      Irving    TX  32.767268  -96.777626        -6   
3  60401     Beecher    IL  41.350484  -87.624080        -6   
4  80432        Como    CO  39.243440 -105.794310        -7   

   Daylight savings time flag              geopoint  
0                           1   38.654948,-99.32062  
1                           0  32.335122,-111.14888  
2                           1  32.767268,-96.777626  
3                           1   41.350484,-87.62408  
4                           1   39.24344,-105.79431

In [2]:
# transform the list into a pandas dataframe
df = pd.DataFrame(list)
df.head()

Zip        City State   Latitude   Longitude  Timezone  \
0  67553  Liebenthal    KS  38.654948  -99.320620        -6   
1  85743      Tucson    AZ  32.335122 -111.148880        -7   
2  75016      Irving    TX  32.767268  -96.777626        -6   
3  60401     Beecher    IL  41.350484  -87.624080        -6   
4  80432        Como    CO  39.243440 -105.794310        -7   

   Daylight savings time flag              geopoint  
0                           1   38.654948,-99.32062  
1                           0  32.335122,-111.14888  
2                           1  32.767268,-96.777626  
3                           1   41.350484,-87.62408  
4                           1   39.24344,-105.79431

In [3]:
# select only zip codes in Austin

df_austin = df[df.City == 'Austin']
df_austin.head()

Zip    City State   Latitude   Longitude  Timezone  \
145   78727  Austin    TX  30.425652  -97.714190        -6   
471   78762  Austin    TX  30.326374  -97.771258        -6   
1271  78763  Austin    TX  30.335398  -97.559807        -6   
1289  78715  Austin    TX  30.450088  -97.486509        -6   
2142  81410  Austin    CO  38.790530 -107.974040        -7   

      Daylight savings time flag              geopoint  
145                            1   30.425652,-97.71419  
471                            1  30.326374,-97.771258  
1271                           1  30.335398,-97.559807  
1289                           1  30.450088,-97.486509  
2142                           1   38.79053,-107.97404

In [4]:
# it looks like there are Austins in multiple states, select only Texas
df_austx = df_austin[df_austin.State == 'TX']
df_austx

Zip    City State   Latitude  Longitude  Timezone  \
145    78727  Austin    TX  30.425652 -97.714190        -6   
471    78762  Austin    TX  30.326374 -97.771258        -6   
1271   78763  Austin    TX  30.335398 -97.559807        -6   
1289   78715  Austin    TX  30.450088 -97.486509        -6   
2144   78745  Austin    TX  30.207559 -97.795750        -6   
2606   78759  Austin    TX  30.406169 -97.757430        -6   
2685   78774  Austin    TX  30.326374 -97.771258        -6   
3617   78701  Austin    TX  30.271270 -97.741030        -6   
4052   78705  Austin    TX  30.292424 -97.738560        -6   
4736   78703  Austin    TX  30.290907 -97.762770        -6   
6571   78754  Austin    TX  30.354234 -97.646790        -6   
6675   78712  Austin    TX  30.285207 -97.735394        -6   
6806   78751  Austin    TX  30.310707 -97.723000        -6   
6953   78752  Austin    TX  30.332506 -97.705710        -6   
8777   78702  Austin    TX  30.265158 -97.718790        -6   
8838   78742  Austin    TX  30.233580 -97.678310        -6   
9056   78786  Austin    TX  30.326374 -97.771258        -6   
10936  78780  Austin    TX  30.326374 -97.771258        -6   
11245  78717  Austin    TX  30.494623 -97.756870        -6   
11607  78771  Austin    TX  30.326374 -97.771258        -6   
11825  78781  Austin    TX  30.326374 -97.771258        -6   
12460  78737  Austin    TX  30.190250 -97.958540        -6   
12615  78728  Austin    TX  30.451803 -97.679890        -6   
13531  78732  Austin    TX  30.382724 -97.894590        -6   
14290  78741  Austin    TX  30.231252 -97.716000        -6   
14502  78767  Austin    TX  30.222007 -97.896285        -6   
14719  78749  Austin    TX  30.216108 -97.858280        -6   
15404  78785  Austin    TX  30.326374 -97.771258        -6   
15443  78735  Austin    TX  30.250761 -97.844690        -6   
16961  73344  Austin    TX  30.326374 -97.771258        -6   
...      ...     ...   ...        ...        ...       ...   
31623  73301  Austin    TX  30.326374 -97.771258        -6   
31690  78724  Austin    TX  30.294148 -97.628630        -6   
32004  78764  Austin    TX  30.445502 -97.659533        -6   
32321  78758  Austin    TX  30.384204 -97.703920        -6   
32622  78769  Austin    TX  30.326374 -97.771258        -6   
33001  78747  Austin    TX  30.132855 -97.761870        -6   
33365  78756  Austin    TX  30.320206 -97.741770        -6   
33441  78736  Austin    TX  30.245558 -97.941770        -6   
33606  78765  Austin    TX  30.326374 -97.771258        -6   
34743  78720  Austin    TX  30.326374 -97.771258        -6   
35131  78704  Austin    TX  30.246309 -97.760870        -6   
35494  78738  Austin    TX  30.340111 -97.988690        -6   
35871  78714  Austin    TX  30.335787 -97.443751        -6   
36085  78722  Austin    TX  30.289307 -97.716590        -6   
37683  78779  Austin    TX  30.326374 -97.771258        -6   
37841  78761  Austin    TX  30.326374 -97.771258        -6   
37960  78746  Austin    TX  30.287739 -97.802200        -6   
38103  78972  Austin    TX  30.315360 -97.663293        -6   
38444  78730  Austin    TX  30.359935 -97.831250        -6   
38859  78766  Austin    TX  30.442202 -97.623330        -6   
39588  78719  Austin    TX  30.163458 -97.677110        -6   
39770  78711  Austin    TX  30.326374 -97.771258        -6   
39771  78726  Austin    TX  30.439053 -97.835030        -6   
40562  78788  Austin    TX  30.326374 -97.771258        -6   
40608  78710  Austin    TX  30.351953 -97.715123        -6   
41461  78750  Austin    TX  30.438933 -97.803830        -6   
42471  78757  Austin    TX  30.349455 -97.733280        -6   
42472  78778  Austin    TX  30.326374 -97.771258        -6   
42755  78755  Austin    TX  30.326374 -97.771258        -6   
42920  78713  Austin    TX  30.468583 -97.843336        -6   

       Daylight savings time flag              geopoint  
145                             1   30.425652,-97.71419  
471                             1  30.326374,-97

In [5]:
# clean up the list and keep only necessary columns

df_aus = df_austx.drop(['Timezone', 'Daylight savings time flag', 'geopoint'], axis=1)

df_aus.head()

Zip    City State   Latitude  Longitude
145   78727  Austin    TX  30.425652 -97.714190
471   78762  Austin    TX  30.326374 -97.771258
1271  78763  Austin    TX  30.335398 -97.559807
1289  78715  Austin    TX  30.450088 -97.486509
2144  78745  Austin    TX  30.207559 -97.795750

In [6]:
# reset the index column
df_auszip = df_aus.reset_index()
df_auszip.head()

index    Zip    City State   Latitude  Longitude
0    145  78727  Austin    TX  30.425652 -97.714190
1    471  78762  Austin    TX  30.326374 -97.771258
2   1271  78763  Austin    TX  30.335398 -97.559807
3   1289  78715  Austin    TX  30.450088 -97.486509
4   2144  78745  Austin    TX  30.207559 -97.795750

In [7]:
# drop the original index column

df_auszip = df_auszip.drop(['index'], axis=1)

df_auszip.head()

Zip    City State   Latitude  Longitude
0  78727  Austin    TX  30.425652 -97.714190
1  78762  Austin    TX  30.326374 -97.771258
2  78763  Austin    TX  30.335398 -97.559807
3  78715  Austin    TX  30.450088 -97.486509
4  78745  Austin    TX  30.207559 -97.795750

In [8]:
# install maps
!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    altair-3.1.0               |           py36_0         724 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be 

In [9]:
# create a map of Austin, TX
latitude = 30.2672
longitude = -97.7431
map_aus = folium.Map(location=[latitude, longitude], zoom_start=10)
map_aus

In [10]:
# put the zip codes on the map
for lat, lng, zipcode in zip(df_auszip['Latitude'], df_auszip['Longitude'], df_auszip['Zip']):
    label = '{}'.format(zipcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_aus)  
map_aus

In [11]:
# Define foursquare credentials
CLIENT_ID = 'LACPSRHKN1XSQ0JW1PXJLUOO5IEKP2FOAIF1BOTQIA3JO5SI' # your Foursquare ID
CLIENT_SECRET = 'HJ4IKEXCT4NQNELBA4QIO1JOOTSKSY3PJS302NNP1DUKB0F2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LACPSRHKN1XSQ0JW1PXJLUOO5IEKP2FOAIF1BOTQIA3JO5SI
CLIENT_SECRET:HJ4IKEXCT4NQNELBA4QIO1JOOTSKSY3PJS302NNP1DUKB0F2


In [12]:
# get the first zip in the dataframe
df_auszip.loc[0, 'Zip']

78727

In [13]:
# get the zip's lat/long values

zip_latitude = df_auszip.loc[0, 'Latitude'] # zip latitude value
zip_longitude = df_auszip.loc[0, 'Longitude'] # zip longitude value

zip_num = df_auszip.loc[0, 'Zip'] # zip num

print('Latitude and longitude values of {} are {}, {}.'.format(zip_num, 
                                                               zip_latitude, 
                                                               zip_longitude))

Latitude and longitude values of 78727 are 30.425652000000003, -97.71419.


In [14]:
# top 100 venues within 500 meters

# limit of number of venues returned by Foursquare API
LIMIT = 500

# define radius - 2 miles from center of zip
radius = 3218

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    zip_latitude, 
    zip_longitude, 
    radius, 
    LIMIT)
# display URL
url 

'https://api.foursquare.com/v2/venues/explore?&client_id=LACPSRHKN1XSQ0JW1PXJLUOO5IEKP2FOAIF1BOTQIA3JO5SI&client_secret=HJ4IKEXCT4NQNELBA4QIO1JOOTSKSY3PJS302NNP1DUKB0F2&v=20180605&ll=30.425652000000003,-97.71419&radius=3218&limit=500'

In [15]:
# send the get request to foursquare
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d3858fc8afbe0002d10d101'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'},
    {'name': '$-$$$$', 'key': 'price'}]},
  'headerLocation': 'Austin',
  'headerFullLocation': 'Austin',
  'headerLocationGranularity': 'city',
  'totalResults': 187,
  'suggestedBounds': {'ne': {'lat': 30.45461402896203,
    'lng': -97.68066525396229},
   'sw': {'lat': 30.396689971037976, 'lng': -97.74771474603772}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '460d9079f964a5200c451fe3',
       'name': 'Tomodachi Sushi',
       'location': {'address': '4101 W Parmer Ln Ste E',
        'crossStreet': 'at Silver Creek Dr',
        'lat': 30.4255931493702,
        'lng': -97.71689383160869,


In [16]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
print("done")

done


In [17]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [18]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d38592c6bdee60039c4ab12'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Austin',
  'headerFullLocation': 'Austin',
  'headerLocationGranularity': 'city',
  'totalResults': 187,
  'suggestedBounds': {'ne': {'lat': 30.45461402896203,
    'lng': -97.68066525396229},
   'sw': {'lat': 30.396689971037976, 'lng': -97.74771474603772}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '460d9079f964a5200c451fe3',
       'name': 'Tomodachi Sushi',
       'location': {'address': '4101 W Parmer Ln Ste E',
        'crossStreet': 'at Silver Creek Dr',
        'lat': 30.4255931493702,
        'lng': -97.71689383160869,


In [19]:
# create a venue table
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0            Tomodachi Sushi      Sushi Restaurant  30.425593 -97.716894
1               Thai Cuisine       Thai Restaurant  30.425621 -97.716874
2  It's A Grind Coffee House           Coffee Shop  30.424973 -97.715844
3          Thundercloud Subs        Sandwich Place  30.421994 -97.702415
4                Chick-fil-A  Fast Food Restaurant  30.419015 -97.703302

In [20]:
# transform into a dataframe
df_venues = pd.DataFrame(nearby_venues)
df_venues.head()

name            categories        lat        lng
0            Tomodachi Sushi      Sushi Restaurant  30.425593 -97.716894
1               Thai Cuisine       Thai Restaurant  30.425621 -97.716874
2  It's A Grind Coffee House           Coffee Shop  30.424973 -97.715844
3          Thundercloud Subs        Sandwich Place  30.421994 -97.702415
4                Chick-fil-A  Fast Food Restaurant  30.419015 -97.703302

In [21]:
# how many breweries are in 78727 zip code?
df_venues[df_venues['categories'].str.contains('brew')]

Empty DataFrame
Columns: [name, categories, lat, lng]
Index: []

## It looks like there aren't any breweries within 2 miles of the center of this zip code. Let's expand the search to all zip codes.

In [22]:
def getNearbyVenues(names, latitudes, longitudes, radius=3218):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [23]:
# create a new dataframe for all Auston venues

aus_ven = getNearbyVenues(names=df_auszip['Zip'],
                                   latitudes=df_auszip['Latitude'],
                                   longitudes=df_auszip['Longitude']
                                  )


78727
78762
78763
78715
78745
78759
78774
78701
78705
78703
78754
78712
78751
78752
78702
78742
78786
78780
78717
78771
78781
78737
78728
78732
78741
78767
78749
78785
78735
73344
78760
78782
78718
78744
78783
78753
78733
78723
78708
78789
78773
78721
78787
78716
78768
78709
78734
78739
78725
78772
78731
78729
78748
73301
78724
78764
78758
78769
78747
78756
78736
78765
78720
78704
78738
78714
78722
78779
78761
78746
78972
78730
78766
78719
78711
78726
78788
78710
78750
78757
78778
78755
78713


In [26]:
print(aus_ven.shape)
aus_ven.head()

(6794, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         78727              30.425652               -97.71419   
1         78727              30.425652               -97.71419   
2         78727              30.425652               -97.71419   
3         78727              30.425652               -97.71419   
4         78727              30.425652               -97.71419   

                       Venue  Venue Latitude  Venue Longitude  \
0            Tomodachi Sushi       30.425593       -97.716894   
1               Thai Cuisine       30.425621       -97.716874   
2  It's A Grind Coffee House       30.424973       -97.715844   
3          Thundercloud Subs       30.421994       -97.702415   
4                Chick-fil-A       30.419015       -97.703302   

         Venue Category  
0      Sushi Restaurant  
1       Thai Restaurant  
2           Coffee Shop  
3        Sandwich Place  
4  Fast Food Restaurant

In [27]:
# clean up the venues table
aus_ven.columns = ['Zip', 'ziplat', 'ziplong', 'venue', 'venlat', 'venlong', 'vencat']
aus_ven.head()

Zip     ziplat   ziplong                      venue     venlat  \
0  78727  30.425652 -97.71419            Tomodachi Sushi  30.425593   
1  78727  30.425652 -97.71419               Thai Cuisine  30.425621   
2  78727  30.425652 -97.71419  It's A Grind Coffee House  30.424973   
3  78727  30.425652 -97.71419          Thundercloud Subs  30.421994   
4  78727  30.425652 -97.71419                Chick-fil-A  30.419015   

     venlong                vencat  
0 -97.716894      Sushi Restaurant  
1 -97.716874       Thai Restaurant  
2 -97.715844           Coffee Shop  
3 -97.702415        Sandwich Place  
4 -97.703302  Fast Food Restaurant

# Methodology - filter the foursquare venues for Breweries and cluster the zip codes by number of breweries. This will identify which zip codes have the most breweries and therefore which zip code I should find a hotel in. 

In [28]:
# filter for breweries
df_brew = aus_ven[aus_ven.vencat == 'Brewery']
df_brew.head()


Zip     ziplat   ziplong                        venue     venlat  \
294   78745  30.207559 -97.79575        (512) Brewing Company  30.222925   
408   78759  30.406169 -97.75743    North By Northwest (NXNW)  30.390949   
688   78705  30.292424 -97.73856  Draught House Pub & Brewery  30.310943   
1023  78751  30.310707 -97.72300               The Brewtorium  30.325771   
1032  78751  30.310707 -97.72300  Draught House Pub & Brewery  30.310943   

        venlong   vencat  
294  -97.770016  Brewery  
408  -97.738573  Brewery  
688  -97.742718  Brewery  
1023 -97.717083  Brewery  
1032 -97.742718  Brewery

In [29]:
print("It looks like we found", df_brew['venue'].count(), "breweries in Austin! That's a lot of potential clients!")

It looks like we found 43 breweries in Austin! That's a lot of potential clients!


In [30]:
# count the breweries by zip code

df_brewzip = df_brew.groupby('Zip').count()
df_brewzip

ziplat  ziplong  venue  venlat  venlong  vencat
Zip                                                   
78702       4        4      4       4        4       4
78705       1        1      1       1        1       1
78710       1        1      1       1        1       1
78721       6        6      6       6        6       6
78722       1        1      1       1        1       1
78723       1        1      1       1        1       1
78732       1        1      1       1        1       1
78736       1        1      1       1        1       1
78741       4        4      4       4        4       4
78742       3        3      3       3        3       3
78744       2        2      2       2        2       2
78745       1        1      1       1        1       1
78746       1        1      1       1        1       1
78751       2        2      2       2        2       2
78752       2        2      2       2        2       2
78756       2        2      2       2        2       2
78757       1        1      1       1        1       1
78758       6        6      6       6        6       6
78759       1        1      1       1        1       1
78764       1        1      1       1        1       1
78972       1        1      1       1        1       1

In [31]:
df_brewzip_count = pd.DataFrame(df_brewzip.vencat)
df_brewzip_count


vencat
Zip          
78702       4
78705       1
78710       1
78721       6
78722       1
78723       1
78732       1
78736       1
78741       4
78742       3
78744       2
78745       1
78746       1
78751       2
78752       2
78756       2
78757       1
78758       6
78759       1
78764       1
78972       1

In [36]:
import matplotlib.pyplot as plt
import pandas as pd

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [38]:
# cluster the breweries

# set number of clusters
kclusters = 5


# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(df_brewzip_count)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 0, 0, 2, 0, 0, 0, 0, 3, 4], dtype=int32)

In [50]:
# create a new df which includes the cluster

# add clustering labels
df_brewzip_count.insert(0, 'Cluster Labels', kmeans.labels_)

ValueError: cannot insert Cluster Labels, already exists

In [57]:
df_brewzip_count.reset_index()

Zip  Cluster Labels  vencat
0   78702               2       4
1   78705               0       1
2   78710               0       1
3   78721               1       6
4   78722               0       1
5   78723               0       1
6   78732               0       1
7   78736               0       1
8   78741               2       4
9   78742               4       3
10  78744               3       2
11  78745               0       1
12  78746               0       1
13  78751               3       2
14  78752               3       2
15  78756               3       2
16  78757               0       1
17  78758               1       6
18  78759               0       1
19  78764               0       1
20  78972               0       1

In [66]:
df_brewzipcount = pd.DataFrame(df_brewzip_count.reset_index())

In [68]:
df_brewzipcount.columns

Index(['Zip', 'Cluster Labels', 'vencat'], dtype='object')

In [72]:
# merge cluster number with zip code lat long df
aus_merged = df_auszip

aus_merged = aus_merged.join(df_brewzipcount.set_index('Zip'), on='Zip')

aus_merged.head()

Zip    City State   Latitude  Longitude  Cluster Labels  vencat
0  78727  Austin    TX  30.425652 -97.714190             NaN     NaN
1  78762  Austin    TX  30.326374 -97.771258             NaN     NaN
2  78763  Austin    TX  30.335398 -97.559807             NaN     NaN
3  78715  Austin    TX  30.450088 -97.486509             NaN     NaN
4  78745  Austin    TX  30.207559 -97.795750             0.0     1.0

In [75]:
aus_merged.dropna(subset=['Cluster Labels'], inplace = True)

In [82]:
aus_merged.astype({'Cluster Labels': 'int64'})

Zip    City State   Latitude  Longitude  Cluster Labels  vencat
4   78745  Austin    TX  30.207559 -97.795750               0     1.0
5   78759  Austin    TX  30.406169 -97.757430               0     1.0
8   78705  Austin    TX  30.292424 -97.738560               0     1.0
12  78751  Austin    TX  30.310707 -97.723000               3     2.0
13  78752  Austin    TX  30.332506 -97.705710               3     2.0
14  78702  Austin    TX  30.265158 -97.718790               2     4.0
15  78742  Austin    TX  30.233580 -97.678310               4     3.0
23  78732  Austin    TX  30.382724 -97.894590               0     1.0
24  78741  Austin    TX  30.231252 -97.716000               2     4.0
33  78744  Austin    TX  30.188377 -97.740380               3     2.0
37  78723  Austin    TX  30.306507 -97.686510               0     1.0
41  78721  Austin    TX  30.272926 -97.686650               1     6.0
55  78764  Austin    TX  30.445502 -97.659533               0     1.0
56  78758  Austin    TX  30.384204 -97.703920               1     6.0
59  78756  Austin    TX  30.320206 -97.741770               3     2.0
60  78736  Austin    TX  30.245558 -97.941770               0     1.0
66  78722  Austin    TX  30.289307 -97.716590               0     1.0
69  78746  Austin    TX  30.287739 -97.802200               0     1.0
70  78972  Austin    TX  30.315360 -97.663293               0     1.0
77  78710  Austin    TX  30.351953 -97.715123               0     1.0
79  78757  Austin    TX  30.349455 -97.733280               0     1.0

In [88]:
# visualize the clusters
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = map_aus

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, zipcode, cluster in zip(aus_merged['Latitude'], aus_merged['Longitude'], aus_merged['Zip'], aus_merged['Cluster Labels']):
    label = folium.Popup(str(zipcode) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)],
        fill=True,
        fill_color=rainbow[int(cluster)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### It looks like zip codes 78721 and 78758 have the most amount of breweries near them, I should stay in a hotel in one of these two zip codes. 

In [91]:
# pull the ratings of these venues

df_hotels = aus_ven[aus_ven.vencat == 'Hotel']
df_hotels.head()

Zip     ziplat    ziplong  \
33   78727  30.425652 -97.714190   
77   78727  30.425652 -97.714190   
86   78727  30.425652 -97.714190   
218  78763  30.335398 -97.559807   
300  78745  30.207559 -97.795750   

                                               venue     venlat    venlong  \
33                               Archer Hotel Austin  30.403281 -97.724080   
77                            Home2 Suites By Hilton  30.399220 -97.718518   
86                        Aloft Austin at The Domain  30.402759 -97.723693   
218  Americas Best Value Inn & Suites Manor Austin E  30.348233 -97.553964   
300        Holiday Inn Express & Suites Austin South  30.199401 -97.767841   

    vencat  
33   Hotel  
77   Hotel  
86   Hotel  
218  Hotel  
300  Hotel

### how many hotels are in zip code 78721?

In [94]:
df_hotels_78721 = df_hotels[df_hotels.Zip == 78721]
print("It looks like we found", df_hotels_78721['venue'].count(), "hotels in 78721")

It looks like we found 0 hotels in 78721


In [95]:
df_hotels_78758 = df_hotels[df_hotels.Zip == 78758]
print("It looks like we found", df_hotels_78758['venue'].count(), "hotels in 78758")

It looks like we found 3 hotels in 78758


In [97]:
# get a list of the hotels
df_hotels_78758

Zip     ziplat   ziplong                   venue     venlat  \
4662  78758  30.384204 -97.70392  Home2 Suites By Hilton  30.399220   
4667  78758  30.384204 -97.70392         Lone Star Court  30.395107   
4708  78758  30.384204 -97.70392     Archer Hotel Austin  30.403281   

        venlong vencat  
4662 -97.718518  Hotel  
4667 -97.726181  Hotel  
4708 -97.724080  Hotel

In [107]:
results

{'meta': {'code': 200, 'requestId': '5d38592c6bdee60039c4ab12'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Austin',
  'headerFullLocation': 'Austin',
  'headerLocationGranularity': 'city',
  'totalResults': 187,
  'suggestedBounds': {'ne': {'lat': 30.45461402896203,
    'lng': -97.68066525396229},
   'sw': {'lat': 30.396689971037976, 'lng': -97.74771474603772}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '460d9079f964a5200c451fe3',
       'name': 'Tomodachi Sushi',
       'location': {'address': '4101 W Parmer Ln Ste E',
        'crossStreet': 'at Silver Creek Dr',
        'lat': 30.4255931493702,
        'lng': -97.71689383160869,


In [111]:
# create a hotel table
hotels = results['response']['groups'][0]['items']
    
nearby_hotels = json_normalize(hotels) # flatten JSON

# filter columns
filtered_columns_hotels = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng', 'venue.id', 'venue.rating']
nearby_hotels =nearby_hotels.loc[:, filtered_columns_hotels]

# filter the category for each row
nearby_hotels['venue.categories'] = nearby_hotels.apply(get_category_type, axis=1)

# clean columns
nearby_hotels.columns = [col.split(".")[-1] for col in nearby_hotels.columns]

nearby_hotels.head()

name            categories        lat        lng  \
0            Tomodachi Sushi      Sushi Restaurant  30.425593 -97.716894   
1               Thai Cuisine       Thai Restaurant  30.425621 -97.716874   
2  It's A Grind Coffee House           Coffee Shop  30.424973 -97.715844   
3          Thundercloud Subs        Sandwich Place  30.421994 -97.702415   
4                Chick-fil-A  Fast Food Restaurant  30.419015 -97.703302   

                         id  rating  
0  460d9079f964a5200c451fe3     NaN  
1  4b465345f964a520aa1d26e3     NaN  
2  4a8da568f964a520611020e3     NaN  
3  4a41141cf964a520bda41fe3     NaN  
4  4b0d6e6ff964a520244823e3     NaN

In [116]:
df_hotels_78758id = df_hotels_78758.join(nearby_hotels.set_index('name'), on='venue')

In [117]:
df_hotels_78758id

Zip     ziplat   ziplong                   venue     venlat  \
4662  78758  30.384204 -97.70392  Home2 Suites By Hilton  30.399220   
4667  78758  30.384204 -97.70392         Lone Star Court  30.395107   
4708  78758  30.384204 -97.70392     Archer Hotel Austin  30.403281   

        venlong vencat categories        lat        lng  \
4662 -97.718518  Hotel      Hotel  30.399220 -97.718518   
4667 -97.726181  Hotel        NaN        NaN        NaN   
4708 -97.724080  Hotel      Hotel  30.403281 -97.724080   

                            id  rating  
4662  555a0f93498e9be2a0588740     NaN  
4667                       NaN     NaN  
4708  56f484d5498ed65c3b06562d     NaN

## we only returned venue IDs for 2 of the hotels, so let's see which one has the highest rating

In [119]:
venue_id_Hilton = '555a0f93498e9be2a0588740' 
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id_Hilton, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/555a0f93498e9be2a0588740?client_id=LACPSRHKN1XSQ0JW1PXJLUOO5IEKP2FOAIF1BOTQIA3JO5SI&client_secret=HJ4IKEXCT4NQNELBA4QIO1JOOTSKSY3PJS302NNP1DUKB0F2&v=20180605'

In [120]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'id': '555a0f93498e9be2a0588740',
 'name': 'Home2 Suites By Hilton',
 'contact': {'phone': '5123392400', 'formattedPhone': '(512) 339-2400'},
 'location': {'address': '2800 Esperanza Xing',
  'lat': 30.399220461627653,
  'lng': -97.71851803802872,
  'labeledLatLngs': [{'label': 'display',
    'lat': 30.399220461627653,
    'lng': -97.71851803802872}],
  'postalCode': '78758',
  'cc': 'US',
  'neighborhood': 'North Burnet',
  'city': 'Austin',
  'state': 'TX',
  'country': 'United States',
  'formattedAddress': ['2800 Esperanza Xing',
   'Austin, TX 78758',
   'United States']},
 'canonicalUrl': 'https://foursquare.com/v/home2-suites-by-hilton/555a0f93498e9be2a0588740',
 'categories': [{'id': '4bf58dd8d48988d1fa931735',
   'name': 'Hotel',
   'pluralName': 'Hotels',
   'shortName': 'Hotel',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/hotel_',
    'suffix': '.png'},
   'primary': True}],
 'verified': False,
 'stats': {'tipCount': 4},
 'url': 'http://home2suites3

In [125]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

8.1


In [127]:
venue_id_Archer = '56f484d5498ed65c3b06562d' 
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id_Archer, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/56f484d5498ed65c3b06562d?client_id=LACPSRHKN1XSQ0JW1PXJLUOO5IEKP2FOAIF1BOTQIA3JO5SI&client_secret=HJ4IKEXCT4NQNELBA4QIO1JOOTSKSY3PJS302NNP1DUKB0F2&v=20180605'

In [128]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'description', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto'])


{'id': '56f484d5498ed65c3b06562d',
 'name': 'Archer Hotel Austin',
 'contact': {'phone': '5128365700',
  'formattedPhone': '(512) 836-5700',
  'twitter': 'archerhotel',
  'facebook': '1472215672995666',
  'facebookUsername': 'ArcherHotelAustin',
  'facebookName': 'Archer Hotel Austin'},
 'location': {'address': '3121 Palm Way',
  'lat': 30.4032805,
  'lng': -97.72407989999999,
  'labeledLatLngs': [{'label': 'display',
    'lat': 30.4032805,
    'lng': -97.72407989999999}],
  'postalCode': '78758',
  'cc': 'US',
  'neighborhood': 'North Burnet',
  'city': 'Austin',
  'state': 'TX',
  'country': 'United States',
  'formattedAddress': ['3121 Palm Way', 'Austin, TX 78758', 'United States']},
 'canonicalUrl': 'https://foursquare.com/v/archer-hotel-austin/56f484d5498ed65c3b06562d',
 'categories': [{'id': '4bf58dd8d48988d1fa931735',
   'name': 'Hotel',
   'pluralName': 'Hotels',
   'shortName': 'Hotel',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/hotel_',
    'suffix'

In [129]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

8.5


# Results - It was determined that the zip codes with the most breweries within 2 miles of the center were 78721 and 78758 through clustering and counting. There were only two hotels within 100 meters of the center of those zip codes, Home2 Suites by hilton and Archer Hotel Austin. The rating of the Home2 Suites By Hilton is 8.1 and the rating for Archer Hotel Austin is 8.5. I will stay at the Archer Hotel Austin, and I hope to have a good sales trip now that I know I'm staying in the best hotel, closest to the most potential clients. 